### DISCUSSION TITLE
IQueryable ProjectTo + other ideas

### DISCUSSION THREAD

**Discussion by CaptTwigg** [riok/mapperly#99] [Ideas]
Hello, thank you for a great tool! 
I'm in the process of replacing AutoMapper with Mapperly, and in that process I have some ideas to new features, that you maybe already have thought about, but anyway.

Not sure if you would like a new discussion for every idea, so I will start with one here.
Generate a "ProjectTo" IQueryable  like [AutoMapper Queryable-Extensions](https://docs.automapper.org/en/stable/Queryable-Extensions.html) and preferable with some recursion with a limit `[RecursionDepth(<some number>)]`?

**Comment by ziaulhasanhamim** [comment]
`ProjectTo` is very essential for ef core mappings. Also there can be a method like `GetMappingExpression` which will return an expression which can be passed into `Select` method of ef core or even mongodb.

**Comment by Thundarr1974** [comment]
Hello. Just wanted to pass along a library I found to accomplish using Mapperly functions in LINQ expressions (for DSL translations like SQL, COSMOS, etc.)

https://github.com/hazzik/DelegateDecompiler

It allows for decompiling the generated function into expression trees. Here's a sample for how we're using it:

```
[Mapper]
public partial class ModelViewMapper
{
    public ModelViewMapper()
    {
        CustomerViewProjection = GetProjection((Func<Customer, CustomerView>)CustomerToCustomerView);
    }

    [MapProperty($"{nameof(Customer.CustomerSubscription)}.{nameof(Customer.CustomerSubscription.Subscription)}.{nameof(Customer.CustomerSubscription.Subscription.Name)}", nameof(CustomerView.Subscription))]
    public partial CustomerView CustomerToCustomerView(Customer customer);

    public Expression<Func<Customer, CustomerView>> CustomerViewProjection { get; }

    private static Expression<Func<TIn, TOut>> GetProjection<TIn, TOut>(Func<TIn, TOut> func) => (Expression<Func<TIn, TOut>>)func.Decompile();
}
....

        var db = ServiceProvider.GetRequiredService<SomeDbContext>();
        var mapper = ServiceProvider.GetRequiredService<ModelViewMapper>();

        return await db.Customers
            .AsNoTracking()
            .Include(c => c.CustomerSubscription.Subscription)
            .Select(mapper.CustomerViewProjection)
            .GetPaged(request, c => c.Id, cancellationToken);

```

**Comment by latonz** [comment]
@CaptTwigg ProjectTo should now be supported in `2.8.0-next.1` (see [next docs](https://next.mapperly.pages.dev/docs/configuration/queryable-projections)) and https://github.com/riok/mapperly/issues/214. Any feedback is welcome 😊 Note that these API's are still in an early preview and could break again until 2.8.0 is released.
Closing for now, as the main issue should be resolved. Feel free to open a new discussion for further improvements (eg. regarding the recursion limit).

A recursion limit is not yet supported, could you add a sample for your use case including how the configuration and the generated code could look like?

**Comment by KuraiAndras** [comment]
Hey, @latonz I just checked out the `2.8.0-next.1` and the [documentation](https://next.mapperly.pages.dev/docs/configuration/queryable-projections/) states, that user implemented mappings are not supported. Are there any plans for supporting those?

**Reply by latonz** [reply]
I created https://github.com/riok/mapperly/issues/338 to track this. I think this would be a great feature for Mapperly, however, the implementation may not be straight forward...

**Reply by KuraiAndras** [reply]
So I have looked at it, and maybe I am not understanding it right, but my mapping will always evaluate on clientside (i am testing this with the EF Core SqLite provider). I have the following entities:
```csharp
public class Project
{
    public Guid Id { get; set; }
    public string Name { get; set; } = default!;
    public List<ProjectPlatform> Platforms { get; set; } = default!;
}

public class ProjectPlatform
{
    public Guid Id { get; set; }
    public Platform PlatformType { get; set; }

    public Guid ProjectId { get; set; }
    public Project Project { get; set; } = default!;
}

public enum Platform
{
    Android,
    IOS,
    Web,
}
```

Then a DTO i want to map to:
```csharp
public class ProjectDto
{
    public Guid Id { get; set; }
    public string Name { get; set; } = default!;
    public List<Platform> Platforms { get; set; } = default!;
}
```

With the following mapper:
```csharp
[Mapper]
public static partial class MapperlyMapper
{
    public static partial IQueryable<ProjectDto> ProjectToProjectDto(this IQueryable<Project> projects);

    public static List<Platform> ProjectPlatformToPlatform(List<ProjectPlatform> projects) => projects. Select(project => project.PlatformType).ToList();
}
```

This is the generated code:
```csharp
#nullable enable
namespace Maplayground
{
    public static partial class MapperlyMapper
    {
        public static partial global::System.Linq.IQueryable<global::Maplayground.ProjectDto> ProjectToProjectDto(this global::System.Linq.IQueryable<global::Maplayground.Project> projects)
        {
#nullable disable
            return System.Linq.Queryable.Select(projects, x => new global::Maplayground.ProjectDto() { Id = x.Id, Name = x.Name, Platforms = ProjectPlatformToPlatform(x.Platforms) });
#nullable enable
        }
    }
}
```
Here when mapping from `ProjectPlatform` to `Platform`, the mapper will call a method which will be evaluated on the client side.

Am I doing something wrong or is this the expected behavior?

**Reply by Yutsa** [reply]
I have the same issue of the mapping being done after the SQL request whereas when I use AutoMapper's `ProjectTo` it is done inside the SQL query.

Any update on this ?

**Reply by latonz** [reply]
Does this only happen if you use user implemented methods?

**Reply by Yutsa** [reply]
I actually mixed things up, my bad. I used to use AutoMapper and HotChocolate together to implement a GraphQL API. Using AutoMapper, it somehow managed to only select the fields that were both in the DTO and in the graphql request.

For instance if I have a DB entity with `firstName`, `lastName`, `age` and `address` for instance and a DTO with only `address` and `age`. If in GraphQL I only ask the `age`, the SQL query will only select the `age`.

With Mapperly the SQL request will select `address` and `age` because they are in the DTO, even if I don't ask for them in graphql.

I don't know how it works exactly in the AutoMapper example but it works for some reason. Here's what the code looks like for my GraphQL query with AutoMapper : 

```csharp
    [UsePaging(MaxPageSize = 1000)]
    [UseProjection]
    [UseFiltering]
    [UseSorting]
    public IQueryable<WorkHeaderDto> GetWorkHeader(
        [Service] WorkService workService,
        [GraphQLName("username")] [GraphQLNonNullType] string username,
        [GraphQLName("startDate")] [GraphQLNonNullType] DateTime startDate,
        [GraphQLName("endDate")] [GraphQLNonNullType] DateTime endDate,
        [GraphQLName("paramEntites")] ParametreRecuperationEntitesConcernees paramEntites,
        [GraphQLName("withParcels")] WorkWithParcels withParcels = WorkWithParcels.All)
    {
        if (paramEntites == null)
        {
            paramEntites = new ParametreRecuperationEntitesConcernees();
        }

        return workService.GetQuery(
                username,
                paramEntites,
                startDate,
                endDate,
                withParcels,
                true)
            .ProjectTo<WorkHeaderDto>(WorkHeaderDtoMapper.Configuration);
    }
```

The same thing but using Mapperly does not produce the same SQL request

```csharp
return workService.GetQuery(
                username,
                paramEntites,
                startDate,
                endDate,
                withParcels,
                true)
            .ProjectToDto();
```

**Reply by Yutsa** [reply]
Okay I'm very sorry, forget everything I've said. The mapping method was incomplete and graphql was requesting a field that wasn't mapped. It does work correctly if I request the right fields !

I'm glad that it works and sorry for the inconvenience.

I have the following mapper : 

```csharp
[Mapper]
public static partial class WorkHeaderDtoMapperly
{

    public static WorkHeaderDto WorkToWorkHeaderDto(this WorkHeader workHeader)
    {
        var dto = workHeader.MapInternal();
        dto.StartDate = workHeader.Works.Min(w => w.StartDate);
        dto.EndDate = workHeader.Works.Min(w => w.EndDate);
        dto.Parcelles = workHeader.Works.SelectMany(w => w.Parcelles).Distinct();
        // dto.Tools = workHeader.Works.SelectMany(w => w.Tools).Select(t => t.Tool).Distinct();
        dto.Persons = workHeader.Works.SelectMany(w => w.Persons).Select(w => w.Person).Distinct();
        dto.TypeTravail = workHeader.Works.First().CategorieTravail.TypeTravail.TypeTravailId;
        dto.Travail = workHeader.Works.First().CategorieTravail.CategorieTravailId;
        dto.CategoryId = workHeader.Works.First().CategorieTravail.Category?.Id;
        dto.WorkArea = workHeader.Works.Sum(w => w.Quantity) ?? 0;
        return dto;
    }
    
    public static partial IQueryable<WorkHeaderDto> ProjectToDto(this IQueryable<WorkHeader> workHeader);
    
    private static partial WorkHeaderDto MapInternal(this WorkHeader workHeader);
}
```

And now I only need a way to reuse the custom mappings inside the `IQueryable` mapping method.

**Reply by latonz** [reply]
How does the generated code look like and how do you expect it to look like?

**Reply by Yutsa** [reply]
Basically I have some custom mapping logic that you can see inside my `WorkToWorkHeaderDto`. I managed to make it work in "classic" mapping using Before/After.

However I need the same mapping done for the `IQueryable`. Unfortunately I haven't found a way to make it work. 

The best I've came up with is this :

```csharp
    public static IQueryable<WorkHeaderDto> ProjectToDtoCustom(this IQueryable<WorkHeader> workHeader)
    {
        if (workHeader == null)
            return default;
        return workHeader.Select(workHeader => new WorkHeaderDto
        {
            StartDate = workHeader.Works.Min(w => w.StartDate),
            EndDate = workHeader.Works.Min(w => w.EndDate),
            Parcelles = workHeader.Works.SelectMany(w => w.Parcelles).Distinct(),
            Persons = workHeader.Works.SelectMany(w => w.Persons).Select(w => w.Person).Distinct(),
            TypeTravail = workHeader.Works.First().CategorieTravail.TypeTravail.TypeTravailId,
            Travail = workHeader.Works.First().CategorieTravail.CategorieTravailId,
            CategoryId = workHeader.Works.First().CategorieTravail.Category.Id,
            WorkArea = workHeader.Works.Sum(w => w.Quantity) ?? 0,
        });
    }
```

But it lacks the mapping done inside the generated `MapInternal`. I'm not an expert on `Expression` and its limitations. I tried calling the `WorkToWorkHeaderDto` method inside the `Select` but it selects all fields if I do that because it doesn't convert to an Expression by LINQ I guess.

**Reply by Yutsa** [reply]
By the way here was the generated code I I tried generating the IQueryable method

```csharp
// <auto-generated />
#nullable enable
namespace ApiIdViniteca.Dto.Mappers
{
    public static partial class WorkHeaderDtoMapperly
    {
        public static partial global::System.Linq.IQueryable<global::ApiIdViniteca.Dto.WorkHeaderDto>? ProjectToDto(this global::System.Linq.IQueryable<global::Viniteca.Core.Metier.WorkHeader>? workHeader)
        {
            if (workHeader == null)
                return default;
#nullable disable
            return System.Linq.Queryable.Select(workHeader, x => x == null ? default : new global::ApiIdViniteca.Dto.WorkHeaderDto()
            {
                Id = x.Id,
                EntityCode = x.EntityCode,
                Comment = x.Comment,
                ModifiedBy = x.ModifiedBy,
            });
#nullable enable
        }

        private static partial global::ApiIdViniteca.Dto.WorkHeaderDto? MapInternal(this global::Viniteca.Core.Metier.WorkHeader? workHeader)
        {
            if (workHeader == null)
                return default;
            var target = new global::ApiIdViniteca.Dto.WorkHeaderDto();
            target.Id = workHeader.Id;
            target.EntityCode = workHeader.EntityCode;
            target.Comment = workHeader.Comment;
            target.ModifiedBy = workHeader.ModifiedBy;
            return target;
        }
    }
}
```

The IQueryable method doesn't call the basic one

**Reply by latonz** [reply]
The basic one is not called because the nullable types differ... Does it work if you remove the nullable annotations on the MapInternal types?
